### ISSUE TITLE
errors thrown in sagas are only logged and never thrown

### ISSUE THREAD

**Issue by epikhighs** [redux-saga/redux-saga#1317]
It seems the default behavior in sagas is to log errors with a console.error instead of actually letting the error bubble out to be an uncaught exception.  Is there a way to let the exception actually get thrown?  

For example, if I try to throw new Error() inside a saga, then I see a console.error appear pointing to utils.js ln 225, which is a log function.

Or are errors always supposed to be handled in sagas and never uncaught?

**Comment by Andarist** [comment]
We are reworking errors at the moment, so this is very much subject to discuss right now. There is no way though for you to "catch" uncaught errors from saga - because they are thrown asynchronously - if you want to really handle them you need to put appropriate try/catches in your sagas.

When creating your middleware you can pass [`onError`](https://redux-saga.js.org/docs/api/#createsagamiddlewareoptions) function, but you can't recover from it - just use it for some logging purposes etc.

Is this what you were looking for? What you'd have expect us to handle differently? What your ideal situation would look like?

**Comment by epikhighs** [comment]
Thanks for the clarification.

The problem I'm trying to solve is to be notified of uncaught exceptions in sagas so that I can properly handle the exception.  I do not have to recover from them.  So I will try adding the onError function to the middleware.

As for the ideal solution, I would think this is problem outside of redux-saga.  In other words, as you pointed out, it's a general problem with error handling async code.  As long as we have a means to be notified of all uncaught exceptions via the middleware onError function, then it feels to me that's the best thing redux-saga can do so that ultimately, the developer can add the proper error handling where needed.

Having said all of this, I would assume most developers (especially those new to redux-saga and async programming) do not fully consider how to properly implement error handling.  So more examples, conventions and recommendations would help everyone in this regard.

**Comment by Andarist** [comment]
Adding useful warnings is definitely a good idea. Will think what we can do about it, although one will get notified only after his/her saga crashes. Definitely need better guides too.

**Comment by AdamPflug** [comment]
So most of my experience with Generators is from [Koa v1](https://github.com/koajs/koa) (which delegated a lot to [co](https://github.com/tj/co)), but I remember exceptions [bubbling up](https://github.com/koajs/koa/wiki/Error-Handling) through the levels of middleware in that case - and it was super super handy. You could catch errors at any level, which let you choose to handle them based on where it made sense for that class of error (and reduced a lot of the boilerplate around handling similar errors). I'd love it if I could do something similar with sagas.

That said, I think they're basically converting all their generators to promises with co.wrap and hooking into the error handling behavior of promises to implement this - not sure that's possible for saga since I haven't dug into the implementation details.

**Comment by Andarist** [comment]
Errors in saga can be caught at any level IF you are "chaining" (nesting) blocking `call`s.

It's not possible if you start using `fork` because it would actually make it really, really weird - by the time inner saga crashes the outer saga might be way past `try/catch` block.

**Comment by ggarek** [comment]
@Andarist I think, it would be useful to add some example on how `onError` is meant to be used. I saw, things like this, for example:

```
const sagaMiddleware = createSagaMiddleware({
  onError: error => {
    sendErrorReport(error);
    throw error;
  }
});
```

In this case, `throw` has no effect at all (it is not even logged to console), but it seems it may be not obvious.

**Comment by adipascu** [comment]
I'd prefer if the error would bubble up and crash the app. I can create a fork of the sagas and catch errors in those if I want to.

**Comment by restrry** [comment]
@adipascu from comment above:
> There is no way though for you to "catch" uncaught errors from saga - because they are thrown asynchronously - if you want to really handle them you need to put appropriate try/catches in your sagas.

so you can catch errors with `onError` hook and do whatever you want - show error message, log error stack, send error stack to any error tracking service etc.

**Comment by Andarist** [comment]
In redux-saga@1 (currently in beta) errors always bubble up (cannot bubble through fork boundaries though - u need to handle them inside the forked task). `onError` can be used for reporting purposes - but this might change a little bit because at the moment we can specify both `logger` and `onError` and it seems that their purpose is pretty much the same.

If this matter is still unclear please open a new issue.

**Comment by burgalon** [comment]
hackish solution:
```
  console.error = (e) => { setTimeout(() => {throw e}, 0) }
```
